### PR TITLE
net: ipv4: Allow IGMP packets with zero src address

### LINKED ISSUE #69917: Don't drop IGMP packets with a source addr 0.0.0.0
## Introduction

In networks that use IGMP, IGMP proxy reports can be sent with a source IP address 0.0.0.0.
This is documented in Section 2.1.1 IGMP Forwarding Rules of RFC 4541 Considerations for IGMP and MLD Snooping Switches.

image

### Problem description

Currently the Zephyr network stack drops packets with a source IP address of 0.0.0.0 unless the destination IP is broadcast (this is to allow DHCP negotiation). https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/ip/ipv4.c#L319

image

If a device using Zephyr tries to use multicast on an IGMP enabled network with a proxy querier, it will be dropped from its multicast groups after its initial membership period expires.

This is because the device drops the incoming general membership query (with source IP address 0.0.0.0) and therefore does not respond with a join report to retain membership in the multicast group.

An example is a network where PTP is used for synchronization. Devices communicate using multicast to keep in synchronization and to elect a clock master. If a device loses membership of the PTP multicast traffic group, it will not be able to participate in the PTP communication and will lose synchronization with the rest of the PTP system.

### Proposed change

Don't drop incoming packets with a source IP address of 0.0.0.0 if the IP protocol is IGMP.
We could also restrict it further and only allow the packet if the destination address is 224.0.0.1 (The IGMP all systems multicast address which is the destination for general membership queries).

### Proposed change (Detailed)

```
diff --git a/subsys/net/ip/ipv4.c b/subsys/net/ip/ipv4.c
index 0392e75cec1..5f1f927a8c9 100644
--- a/subsys/net/ip/ipv4.c
+++ b/subsys/net/ip/ipv4.c
@@ -291,7 +291,8 @@ enum net_verdict net_ipv4_input(struct net_pkt *pkt)
        }
 
        if (net_ipv4_is_addr_unspecified((struct in_addr *)hdr->src) &&
-           !net_ipv4_is_addr_bcast(net_pkt_iface(pkt), (struct in_addr *)hdr->dst)) {
+           !net_ipv4_is_addr_bcast(net_pkt_iface(pkt), (struct in_addr *)hdr->dst) &&
+               (hdr->proto != IPPROTO_IGMP)) {
                NET_DBG("DROP: src addr is %s", "unspecified");
                goto drop;
        }

```
### Dependencies

Highlight how the change may affect the rest of the project (new components,
modifications in other areas), or other teams/projects.

### Concerns and Unresolved Questions

List any concerns, unknowns, and generally unresolved questions etc.

## Alternatives

List any alternatives considered, and the reasons for choosing this option
over them.

### PR DESCRIPTION
**Pull request by ad8sweera**
IGMP queries sent out by a proxy querier can have
a source IP address 0.0.0.0
Allow these incoming packets.

Fixes #69917

### CODE REVIEW

File: subsys/net/ip/ipv4.c
Comment (jukkar) [inline subsys/net/ip/ipv4.c:317]: Please change the commit subject to `net: ipv4: Allow IGMP packets with zero src address`
and also add `Fixes #69917` before the signed-off-by line so that the commit is linked to the issue.

File: subsys/net/ip/ipv4.c
Comment (jukkar) [inline subsys/net/ip/ipv4.c:321]: Could you align the added line to same column as the previous line.

**Review by jukkar — COMMENTED** [review]
Looks good, couple of minor nits.

**Review by rlubos — COMMENTED** [review]
Please squash the commits

**Review by jukkar — APPROVED** [review]

**Review by pdgendt — APPROVED** [review]

**Comment by ad8sweera** [comment]
Thanks all, will someone do the merge? (I don't have permission)

**Comment by pdgendt** [comment]
> Thanks all, will someone do the merge? (I don't have permission)

The merge will be done by administrators, but it needs approval from the assignee first, so please wait for a review.

**Review by rlubos — APPROVED** [review]

**Pull request merged**
